torchinductor/scheduler.py

```diff
@@ -763,13 +763,38 @@ def get_possible_fusions(self):
         Helper to find all legal fusion opportunities, sorted by self.score_fusion()
         """
         possible_fusions = []
-        for node1_index, node1 in enumerate(self.nodes):
-            for node2 in self.nodes[node1_index + 1 :]:
-                if self.can_fuse(node1, node2):
-                    possible_fusions.append((node1, node2))
-                elif self.can_fuse(node2, node1):
-                    # epilogue fusions are order dependent
-                    possible_fusions.append((node2, node1))
+        seen = set()
+
+        def check_all_pairs(nodes):
+            for node1_index, node1 in enumerate(nodes):
+                for node2 in nodes[node1_index + 1 :]:
+                    key = (node1, node2)
+                    if key in seen:
+                        continue
+                    seen.add(key)
+
+                    if self.can_fuse(node1, node2):
+                        possible_fusions.append(key)
+                    elif node2.is_template() and self.can_fuse(node2, node1):
+                        # epilogue fusions are order dependent
+                        possible_fusions.append((node2, node1))
+
+        buffer_names_grouping = collections.defaultdict(list)
+        for node in self.nodes:
```

Comment:
Doesn't this miss vertical fusions?

Reply:
used_buffer_names includes writes, so it should get them